unstructured/embed/embedder/open_ai.py

```diff
@@ -0,0 +1,155 @@
+        return OpenAISessionHandle(service=service)
+
+
+class ElementEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, Element):
+            return {
+                "text": obj.text,
+                "element_id": obj.id,
+                "embedding": obj.embedding,
+                "metadata": obj.metadata.to_dict(),
+            }
+        return super().default(obj)
+
+
+@dataclass
+class OpenAIEmbeddingsDoc(IngestDocSessionHandleMixin, DataClassJsonMixin):
+    """Class encapsulating reading elements in a doc and writing the obtained embeddings."""
+
+    config: OpenAIEmbedderConfig
+    filename: str
+
+    @property
+    def _output_filename(self):
+        """Create output file path."""
+        output_file = f"{self.filename}_with_embeddings.json"
+
+        return (Path(self.config.output_dir) / output_file).resolve()
+
+    def get_elements_from_json(self) -> List[Element]:
+        """Reads elements"""
+        return partition_json(filename=self.filename)
+
+    @requires_dependencies(["langchain"], extras="jira")
+    def embed_and_write_result(self):
+        logger.debug(f"Reading: {self} - PID: {os.getpid()}")
+
+        elements = self.get_elements_from_json()
+        embeddings = self.session_handle.service.embed_documents([str(e) for e in elements])
+        self.elements = self.add_embeddings_to_elements(elements, embeddings)
+        # import pdb; pdb.set_trace()
+        self.write_result()
+
+    def add_embeddings_to_elements(self, elements, embeddings):
+        assert len(elements) == len(embeddings)
+
+        for i in range(len(elements)):
+            elements[i].embedding = embeddings[i]
+
+        return elements
+
+    def write_result(self):
+        """Write the structured json result for this doc. result must be json serializable."""
+        self._output_filename.parent.mkdir(parents=True, exist_ok=True)
+        with open(self._output_filename, "w", encoding="utf8") as output_f:
+            json.dump(self.elements, output_f, ensure_ascii=False, indent=2, cls=ElementEncoder)
+        print(f"Wrote {self._output_filename}")
+        logger.info(f"Wrote {self._output_filename}")
+
+
+class OpenAIEmbedder:
+    config: OpenAIEmbedderConfig
+
+    def __init__(
+        self,
+        config: OpenAIEmbedderConfig,
+    ):
+        self.config = config
+
+    @requires_dependencies(["langchain"])  # add extras="langchain" when it's added to the makefile
+    def initialize(self):
+        """Verifies that a connection can be established to OpenAI.
+        Sends one embedding request for verification."""
+        _ = self.config.create_session_handle().service
+
+    def get_embed_docs(self):
+        """Reads all result files to embed them."""
+        return [
+            OpenAIEmbeddingsDoc(
+                self.config,
+                filename,
+            )
+            for filename in self.config.list_of_elements_json_paths.split()
+        ]
```

Comment:
probably for v0 this just takes a single result (list of Elements) or json file and returns the list of Elements with embeddings? i.e. not sure we need to support iterating a list of docs (at least for v0)

Reply:
though maybe we need to look at how we want this to slot in for [ingest here](https://github.com/Unstructured-IO/unstructured/blob/2b571eb9a370a0ac10e3d78815c626015ffb7691/unstructured/ingest/processor.py#L96-L98) for a requirements perspective.

Reply:
Re-implemented as taking a single list of elements and returning the updated list with embeddings in https://github.com/Unstructured-IO/unstructured/pull/1368/commits/499795ae9364593a4954a30e0e568598f3d15f33

Reply:
For fitting into ingest, and a single doc, I think instantiating the class and putting it as `self.embedding_encoder.embed(doc)` should be possible within those lines:
https://github.com/Unstructured-IO/unstructured/blob/2b571eb9a370a0ac10e3d78815c626015ffb7691/unstructured/ingest/processor.py#L96-L98

However, to process multiple docs, we need an async / parallelised solution. Would pool.map be possible like with partition, I wonder:
https://github.com/Unstructured-IO/unstructured/blob/2b571eb9a370a0ac10e3d78815c626015ffb7691/unstructured/ingest/processor.py#L71-L87

A note, since these are API calls and not actual compute intensive operations, our users will ultimately be bound by the model provider (in the current case OpenAI) API. Paid users get good [rate limits](https://platform.openai.com/docs/guides/rate-limits/what-are-the-rate-limits-for-our-api), therefore if OpenAI doesn't have hidden limits behind the scenes (ie if they allocate more resources to the user when the load from that user is high), all types of parallelisation should benefit our users' runtime.